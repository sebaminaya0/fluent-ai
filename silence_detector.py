#!/usr/bin/env python3
"""
Silence Detection Module for Fluent AI

This module provides robust silence/pause detection capabilities using both
webrtcvad and pydub.silence for real-time audio energy monitoring.

Features:
- Real-time audio energy monitoring
- Configurable silence parameters (min_silence_len, silence_thresh)
- Auto-stop transcription when sustained silence exceeds threshold
- Timer reset when speech resumes
- Support for both webrtcvad and pydub silence detection methods
"""

import time
import threading
import logging
import numpy as np
from typing import Optional, Callable, Dict, Any
from collections import deque
import io

try:
    import webrtcvad
    WEBRTCVAD_AVAILABLE = True
except ImportError:
    WEBRTCVAD_AVAILABLE = False
    logging.warning("webrtcvad not available, falling back to pydub-only silence detection")

try:
    from pydub import AudioSegment
    from pydub.silence import detect_nonsilent
    PYDUB_AVAILABLE = True
except ImportError:
    PYDUB_AVAILABLE = False
    logging.warning("pydub not available, falling back to webrtcvad-only silence detection")

import speech_recognition as sr


class SilenceDetector:
    """
    A comprehensive silence detection system that monitors audio energy in real-time
    and provides callbacks for silence/speech events.
    """
    
    def __init__(self, 
                 min_silence_len: int = 800,  # ms
                 silence_thresh: int = -40,   # dBFS
                 frame_duration: int = 30,    # ms (10, 20, or 30 for webrtcvad)
                 aggressiveness: int = 2,     # 0-3 for webrtcvad
                 sample_rate: int = 16000,    # Optimized for Whisper default
                 method: str = 'auto',        # 'auto', 'webrtcvad', 'pydub'
                 chunk_size: int = 1024):     # Optimized chunk size for better performance
        """
        Initialize the silence detector.
        
        Args:
            min_silence_len: Minimum silence duration in ms to trigger silence event
            silence_thresh: Silence threshold in dBFS (negative values)
            frame_duration: Frame duration for webrtcvad (10, 20, or 30 ms)
            aggressiveness: WebRTC VAD aggressiveness (0-3, higher = more aggressive)
            sample_rate: Audio sample rate (8000, 16000, 32000, or 48000 for webrtcvad)
            method: Detection method ('auto', 'webrtcvad', 'pydub')
        """
        self.min_silence_len = min_silence_len
        self.silence_thresh = silence_thresh
        self.frame_duration = frame_duration
        self.aggressiveness = aggressiveness
        self.sample_rate = sample_rate
        self.method = method
        
        # State management
        self.is_monitoring = False
        self.silence_start_time = None
        self.last_speech_time = None
        self.audio_buffer = deque(maxlen=100)  # Keep last 100 frames
        
        # Callbacks
        self.on_silence_detected = None
        self.on_speech_detected = None
        self.on_silence_threshold_exceeded = None
        
        # Initialize detection method
        self._init_detection_method()
        
        # Monitoring thread
        self._monitor_thread = None
        self._stop_monitoring = threading.Event()
        
        logging.info(f"SilenceDetector initialized with method: {self.active_method}")
    
    def _init_detection_method(self):
        """Initialize the detection method based on availability and preferences."""
        if self.method == 'auto':
            if WEBRTCVAD_AVAILABLE:
                self.active_method = 'webrtcvad'
            elif PYDUB_AVAILABLE:
                self.active_method = 'pydub'
            else:
                raise RuntimeError("Neither webrtcvad nor pydub is available")
        elif self.method == 'webrtcvad':
            if not WEBRTCVAD_AVAILABLE:
                raise RuntimeError("webrtcvad is not available")
            self.active_method = 'webrtcvad'
        elif self.method == 'pydub':
            if not PYDUB_AVAILABLE:
                raise RuntimeError("pydub is not available")
            self.active_method = 'pydub'
        else:
            raise ValueError(f"Unknown detection method: {self.method}")
        
        # Initialize WebRTC VAD if using it
        if self.active_method == 'webrtcvad':
            self.vad = webrtcvad.Vad(self.aggressiveness)
            # Validate sample rate for webrtcvad
            if self.sample_rate not in [8000, 16000, 32000, 48000]:
                logging.warning(f"Sample rate {self.sample_rate} not supported by webrtcvad, using 16000")
                self.sample_rate = 16000
            # Validate frame duration
            if self.frame_duration not in [10, 20, 30]:
                logging.warning(f"Frame duration {self.frame_duration} not supported by webrtcvad, using 30")
                self.frame_duration = 30
    
    def set_callbacks(self, 
                     on_silence_detected: Optional[Callable] = None,
                     on_speech_detected: Optional[Callable] = None,
                     on_silence_threshold_exceeded: Optional[Callable] = None):
        """Set callback functions for silence/speech events."""
        self.on_silence_detected = on_silence_detected
        self.on_speech_detected = on_speech_detected
        self.on_silence_threshold_exceeded = on_silence_threshold_exceeded
    
    def update_parameters(self, **kwargs):
        """Update silence detection parameters."""
        for key, value in kwargs.items():
            if hasattr(self, key):
                setattr(self, key, value)
                logging.info(f"Updated {key} to {value}")
        
        # Reinitialize if method-specific parameters changed
        if any(key in ['aggressiveness', 'sample_rate', 'frame_duration'] for key in kwargs):
            if self.active_method == 'webrtcvad':
                self._init_detection_method()
    
    def is_silence_webrtcvad(self, audio_data: bytes) -> bool:
        """Detect silence using WebRTC VAD."""
        try:
            # WebRTC VAD expects 16-bit PCM audio
            is_speech = self.vad.is_speech(audio_data, self.sample_rate)
            return not is_speech
        except Exception as e:
            logging.error(f"WebRTC VAD error: {e}")
            return False
    
    def is_silence_pydub(self, audio_data: bytes) -> bool:
        """Detect silence using pydub."""
        try:
            # Convert bytes to AudioSegment
            audio_segment = AudioSegment(
                audio_data,
                frame_rate=self.sample_rate,
                sample_width=2,  # 16-bit
                channels=1
            )
            
            # Check if the audio level is below threshold
            if audio_segment.dBFS < self.silence_thresh:
                return True
            
            # Also use pydub's silence detection
            nonsilent_ranges = detect_nonsilent(
                audio_segment,
                min_silence_len=self.min_silence_len,
                silence_thresh=self.silence_thresh
            )
            
            return len(nonsilent_ranges) == 0
        except Exception as e:
            logging.error(f"Pydub silence detection error: {e}")
            return False
    
    def detect_silence(self, audio_data: bytes) -> bool:
        """Detect silence using the active method."""
        if self.active_method == 'webrtcvad':
            return self.is_silence_webrtcvad(audio_data)
        elif self.active_method == 'pydub':
            return self.is_silence_pydub(audio_data)
        else:
            return False
    
    def process_audio_frame(self, audio_data: bytes) -> Dict[str, Any]:
        """
        Process a single audio frame and return detection results.
        
        Returns:
            Dictionary containing detection results and timing information
        """
        current_time = time.time()
        is_silent = self.detect_silence(audio_data)
        
        # Add to buffer
        self.audio_buffer.append({
            'timestamp': current_time,
            'is_silent': is_silent,
            'data': audio_data
        })
        
        result = {
            'is_silent': is_silent,
            'timestamp': current_time,
            'silence_duration': 0,
            'speech_detected': False,
            'silence_threshold_exceeded': False
        }
        
        if is_silent:
            if self.silence_start_time is None:
                self.silence_start_time = current_time
                if self.on_silence_detected:
                    self.on_silence_detected(current_time)
            
            # Calculate silence duration
            silence_duration_ms = (current_time - self.silence_start_time) * 1000
            result['silence_duration'] = silence_duration_ms
            
            # Check if silence threshold is exceeded
            if silence_duration_ms >= self.min_silence_len:
                result['silence_threshold_exceeded'] = True
                if self.on_silence_threshold_exceeded:
                    self.on_silence_threshold_exceeded(silence_duration_ms)
        
        else:  # Speech detected
            if self.silence_start_time is not None:
                # Speech resumed, reset silence timer
                self.silence_start_time = None
                result['speech_detected'] = True
                if self.on_speech_detected:
                    self.on_speech_detected(current_time)
            
            self.last_speech_time = current_time
        
        return result
    
    def start_monitoring(self, audio_source):
        """Start monitoring audio source for silence/speech."""
        if self.is_monitoring:
            logging.warning("Silence monitoring already active")
            return
        
        self.is_monitoring = True
        self._stop_monitoring.clear()
        
        def monitor_loop():
            """Main monitoring loop."""
            logging.info("Starting silence monitoring loop")
            
            while not self._stop_monitoring.is_set():
                try:
                    # This would need to be adapted based on the audio source type
                    # For now, this is a placeholder for the monitoring logic
                    time.sleep(0.01)  # Small delay to prevent excessive CPU usage
                except Exception as e:
                    logging.error(f"Error in monitoring loop: {e}")
                    break
            
            logging.info("Silence monitoring loop stopped")
        
        self._monitor_thread = threading.Thread(target=monitor_loop, daemon=True)
        self._monitor_thread.start()
    
    def stop_monitoring(self):
        """Stop monitoring audio source."""
        if not self.is_monitoring:
            return
        
        self.is_monitoring = False
        self._stop_monitoring.set()
        
        if self._monitor_thread and self._monitor_thread.is_alive():
            self._monitor_thread.join(timeout=1.0)
        
        logging.info("Silence monitoring stopped")
    
    def reset_state(self):
        """Reset the detector state."""
        self.silence_start_time = None
        self.last_speech_time = None
        self.audio_buffer.clear()
        logging.info("Silence detector state reset")
    
    def get_stats(self) -> Dict[str, Any]:
        """Get current statistics."""
        current_time = time.time()
        
        silence_duration = 0
        if self.silence_start_time is not None:
            silence_duration = (current_time - self.silence_start_time) * 1000
        
        time_since_speech = None
        if self.last_speech_time is not None:
            time_since_speech = (current_time - self.last_speech_time) * 1000
        
        return {
            'is_monitoring': self.is_monitoring,
            'current_silence_duration': silence_duration,
            'time_since_last_speech': time_since_speech,
            'active_method': self.active_method,
            'min_silence_len': self.min_silence_len,
            'silence_thresh': self.silence_thresh,
            'buffer_size': len(self.audio_buffer)
        }


class SilenceDetectorIntegration:
    """Integration helper for speech recognition systems."""
    
    def __init__(self, recognizer: sr.Recognizer, detector: SilenceDetector):
        """
        Initialize the integration.
        
        Args:
            recognizer: SpeechRecognition recognizer instance
            detector: SilenceDetector instance
        """
        self.recognizer = recognizer
        self.detector = detector
        self.should_stop_listening = False
        self.transcription_callback = None
        
        # Set up detector callbacks
        self.detector.set_callbacks(
            on_silence_threshold_exceeded=self._on_silence_threshold_exceeded
        )
    
    def set_transcription_callback(self, callback: Callable):
        """Set callback for when transcription should be finalized."""
        self.transcription_callback = callback
    
    def _on_silence_threshold_exceeded(self, silence_duration_ms: float):
        """Handle silence threshold exceeded event."""
        logging.info(f"Silence threshold exceeded: {silence_duration_ms:.0f}ms")
        self.should_stop_listening = True
        
        if self.transcription_callback:
            self.transcription_callback(silence_duration_ms)
    
    def listen_with_silence_detection(self, 
                                    source: sr.AudioSource,
                                    timeout: Optional[float] = None,
                                    phrase_time_limit: Optional[float] = None) -> Optional[sr.AudioData]:
        """
        Listen for audio with integrated silence detection.
        
        This method extends the standard speech recognition listening
        with silence detection capabilities.
        """
        self.should_stop_listening = False
        self.detector.reset_state()
        
        try:
            # Custom listening logic that integrates with silence detection
            # This is a simplified version - full implementation would require
            # more sophisticated audio stream processing
            
            audio = self.recognizer.listen(
                source,
                timeout=timeout,
                phrase_time_limit=phrase_time_limit
            )
            
            return audio
            
        except sr.WaitTimeoutError:
            logging.info("Listening timeout reached")
            return None
        except Exception as e:
            logging.error(f"Error during listening: {e}")
            return None
    
    def process_audio_stream(self, audio_stream):
        """Process an audio stream with silence detection."""
        # This method would process an audio stream chunk by chunk
        # and apply silence detection to each chunk
        pass


# Example usage and configuration presets
SILENCE_DETECTION_PRESETS = {
    'sensitive': {
        'min_silence_len': 600,
        'silence_thresh': -35,
        'aggressiveness': 1
    },
    'balanced': {
        'min_silence_len': 800,
        'silence_thresh': -40,
        'aggressiveness': 2
    },
    'aggressive': {
        'min_silence_len': 1200,
        'silence_thresh': -45,
        'aggressiveness': 3
    },
    'very_aggressive': {
        'min_silence_len': 1500,
        'silence_thresh': -50,
        'aggressiveness': 3
    }
}


def create_silence_detector(preset: str = 'balanced', **kwargs) -> SilenceDetector:
    """
    Create a silence detector with a preset configuration.
    
    Args:
        preset: Preset name ('sensitive', 'balanced', 'aggressive', 'very_aggressive')
        **kwargs: Additional parameters to override preset values
    
    Returns:
        Configured SilenceDetector instance
    """
    if preset not in SILENCE_DETECTION_PRESETS:
        raise ValueError(f"Unknown preset: {preset}. Available: {list(SILENCE_DETECTION_PRESETS.keys())}")
    
    config = SILENCE_DETECTION_PRESETS[preset].copy()
    config.update(kwargs)
    
    return SilenceDetector(**config)


if __name__ == "__main__":
    # Example usage
    import argparse
    
    parser = argparse.ArgumentParser(description="Test silence detection")
    parser.add_argument('--preset', default='balanced', 
                       choices=list(SILENCE_DETECTION_PRESETS.keys()),
                       help='Silence detection preset')
    parser.add_argument('--min-silence-len', type=int, default=None,
                       help='Minimum silence length in ms')
    parser.add_argument('--silence-thresh', type=int, default=None,
                       help='Silence threshold in dBFS')
    parser.add_argument('--method', default='auto',
                       choices=['auto', 'webrtcvad', 'pydub'],
                       help='Detection method')
    
    args = parser.parse_args()
    
    # Create detector with preset and overrides
    kwargs = {}
    if args.min_silence_len is not None:
        kwargs['min_silence_len'] = args.min_silence_len
    if args.silence_thresh is not None:
        kwargs['silence_thresh'] = args.silence_thresh
    kwargs['method'] = args.method
    
    detector = create_silence_detector(args.preset, **kwargs)
    
    print(f"Silence detector created with preset: {args.preset}")
    print(f"Configuration: {detector.get_stats()}")
    
    # Example of how to use the detector
    def on_silence_detected(timestamp):
        print(f"Silence detected at {timestamp}")
    
    def on_speech_detected(timestamp):
        print(f"Speech detected at {timestamp}")
    
    def on_silence_threshold_exceeded(duration_ms):
        print(f"Silence threshold exceeded: {duration_ms:.0f}ms")
    
    detector.set_callbacks(
        on_silence_detected=on_silence_detected,
        on_speech_detected=on_speech_detected,
        on_silence_threshold_exceeded=on_silence_threshold_exceeded
    )
    
    print("Silence detector ready for use")
