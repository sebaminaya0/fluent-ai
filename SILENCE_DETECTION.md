# Silence Detection Features

This document describes the robust silence/pause detection system implemented in Fluent AI. The system provides real-time audio energy monitoring with configurable parameters to automatically detect when to stop transcription based on sustained silence.

## Features

### Core Capabilities
- **Real-time audio energy monitoring** using webrtcvad and pydub
- **Configurable silence parameters** (minimum silence duration, audio threshold)
- **Auto-stop transcription** when sustained silence exceeds threshold
- **Timer reset** when speech resumes
- **Multiple detection methods** (webrtcvad, pydub, auto-selection)
- **Preset configurations** for different environments
- **GUI controls** for easy configuration
- **CLI flags** for power users

### Detection Methods

#### WebRTC VAD (Voice Activity Detection)
- **Recommended for:** Real-time speech detection
- **Advantages:** Low latency, optimized for speech
- **Configuration:** Aggressiveness levels 0-3
- **Requirements:** `webrtcvad` package

#### Pydub Silence Detection
- **Recommended for:** General audio analysis
- **Advantages:** Works with any audio content
- **Configuration:** dBFS threshold-based
- **Requirements:** `pydub` package

#### Auto-Selection
- **Default behavior:** Automatically chooses the best available method
- **Fallback order:** webrtcvad â†’ pydub â†’ error

## Configuration

### Silence Parameters

| Parameter | Description | Default | Range |
|-----------|-------------|---------|--------|
| `min_silence_len` | Minimum silence duration to trigger auto-stop | 800ms | 200-2000ms |
| `silence_thresh` | Audio level threshold for silence detection | -40dBFS | -60 to -20dBFS |
| `aggressiveness` | WebRTC VAD aggressiveness (0=least, 3=most) | 2 | 0-3 |
| `method` | Detection method to use | 'auto' | 'auto', 'webrtcvad', 'pydub' |

### Preset Configurations

| Preset | Min Silence | Threshold | Aggressiveness | Use Case |
|--------|-------------|-----------|---------------|----------|
| `sensitive` | 600ms | -35dBFS | 1 | Quiet environments |
| `balanced` | 800ms | -40dBFS | 2 | General use (default) |
| `aggressive` | 1200ms | -45dBFS | 3 | Noisy environments |
| `very_aggressive` | 1500ms | -50dBFS | 3 | Very noisy environments |

## Usage

### CLI Usage

#### Basic Usage
```bash
# Enable silence detection with default settings
python main_whisper.py --silence-detection

# Use a specific preset
python main_whisper.py --silence-detection --silence-preset aggressive

# Custom parameters
python main_whisper.py --silence-detection --min-silence-len 1000 --silence-thresh -35
```

#### Advanced Usage
```bash
# All parameters combined
python main_whisper.py --silence-detection \
  --silence-preset balanced \
  --min-silence-len 1200 \
  --silence-thresh -45 \
  --silence-method webrtcvad \
  --vad-aggressiveness 3
```

### GUI Usage

1. **Enable Detection**: Check the "ðŸ”‡ DetecciÃ³n de silencio automÃ¡tica" checkbox
2. **Choose Preset**: Select from the dropdown (sensitive, balanced, aggressive, very_aggressive)
3. **Fine-tune Parameters**: Use the sliders to adjust:
   - Silence duration (200-2000ms)
   - Audio threshold (-60 to -20dBFS)
4. **Monitor Status**: The status bar shows real-time detection events

### Programmatic Usage

#### Basic Setup
```python
from silence_detector import create_silence_detector, SilenceDetectorIntegration
import speech_recognition as sr

# Create detector with preset
detector = create_silence_detector(preset='balanced')

# Set up callbacks
def on_silence_detected(timestamp):
    print(f"Silence detected at {timestamp}")

def on_speech_detected(timestamp):
    print(f"Speech detected at {timestamp}")

def on_silence_threshold_exceeded(duration_ms):
    print(f"Auto-stop triggered after {duration_ms:.0f}ms of silence")

detector.set_callbacks(
    on_silence_detected=on_silence_detected,
    on_speech_detected=on_speech_detected,
    on_silence_threshold_exceeded=on_silence_threshold_exceeded
)
```

#### Integration with Speech Recognition
```python
# Create components
recognizer = sr.Recognizer()
detector = create_silence_detector(preset='balanced')
integration = SilenceDetectorIntegration(recognizer, detector)

# Set up auto-stop callback
def on_auto_stop(duration_ms):
    print(f"Auto-stopping after {duration_ms:.0f}ms of silence")
    # Implement your auto-stop logic here

integration.set_transcription_callback(on_auto_stop)

# Use with microphone
with sr.Microphone() as source:
    audio = integration.listen_with_silence_detection(source, timeout=30)
    if audio:
        # Process audio with Whisper or other transcription
        pass
```

#### Custom Configuration
```python
# For noisy environments
noisy_detector = create_silence_detector(
    min_silence_len=1200,  # Longer silence required
    silence_thresh=-50,    # Lower threshold for noisy environments
    method='webrtcvad',    # More robust for speech detection
    aggressiveness=3       # Most aggressive VAD setting
)

# For quiet environments
quiet_detector = create_silence_detector(
    min_silence_len=600,   # Shorter silence acceptable
    silence_thresh=-30,    # Higher threshold for quiet environments
    method='pydub',        # Good for general audio analysis
    aggressiveness=1       # Less aggressive VAD setting
)
```

## Installation

### Dependencies
```bash
pip install webrtcvad pydub
```

### Additional Requirements
The silence detection system requires the following packages to be installed:
- `webrtcvad>=2.0.10` - For WebRTC VAD support
- `pydub>=0.25.1` - For general audio silence detection
- `numpy` - For audio processing
- `librosa` - For audio file handling
- `soundfile` - For audio file I/O

These are automatically included in the project's `pyproject.toml` file.

## API Reference

### SilenceDetector Class

#### Constructor
```python
SilenceDetector(
    min_silence_len: int = 800,    # Minimum silence duration in ms
    silence_thresh: int = -40,     # Silence threshold in dBFS
    frame_duration: int = 30,      # Frame duration for webrtcvad (10, 20, or 30 ms)
    aggressiveness: int = 2,       # WebRTC VAD aggressiveness (0-3)
    sample_rate: int = 16000,      # Audio sample rate
    method: str = 'auto'           # Detection method
)
```

#### Methods

| Method | Description |
|--------|-------------|
| `set_callbacks(on_silence_detected, on_speech_detected, on_silence_threshold_exceeded)` | Set event callbacks |
| `update_parameters(**kwargs)` | Update detection parameters |
| `process_audio_frame(audio_data: bytes)` | Process a single audio frame |
| `start_monitoring(audio_source)` | Start monitoring audio source |
| `stop_monitoring()` | Stop monitoring |
| `reset_state()` | Reset detector state |
| `get_stats()` | Get current statistics |

### SilenceDetectorIntegration Class

#### Constructor
```python
SilenceDetectorIntegration(recognizer: sr.Recognizer, detector: SilenceDetector)
```

#### Methods

| Method | Description |
|--------|-------------|
| `set_transcription_callback(callback)` | Set callback for transcription events |
| `listen_with_silence_detection(source, timeout, phrase_time_limit)` | Listen with silence detection |

### Utility Functions

| Function | Description |
|----------|-------------|
| `create_silence_detector(preset='balanced', **kwargs)` | Create detector with preset configuration |

## Examples

### Example 1: Basic CLI Usage
```bash
# Start with silence detection enabled
python main_whisper.py --silence-detection

# Use aggressive preset for noisy environment
python main_whisper.py --silence-detection --silence-preset aggressive
```

### Example 2: GUI Usage
```bash
# Start GUI application
python gui_app.py

# In the GUI:
# 1. Check "ðŸ”‡ DetecciÃ³n de silencio automÃ¡tica"
# 2. Select preset from dropdown
# 3. Adjust sliders as needed
# 4. Start recording - silence detection will be active
```

### Example 3: Custom Integration
```python
import speech_recognition as sr
from silence_detector import create_silence_detector

# Create custom detector
detector = create_silence_detector(
    min_silence_len=1000,
    silence_thresh=-35,
    method='webrtcvad'
)

# Set up callbacks
def handle_silence_threshold(duration_ms):
    print(f"Stopping recording after {duration_ms}ms of silence")
    # Your auto-stop logic here

detector.set_callbacks(
    on_silence_threshold_exceeded=handle_silence_threshold
)

# Use in your application
recognizer = sr.Recognizer()
with sr.Microphone() as source:
    # Your audio processing logic with silence detection
    pass
```

## Troubleshooting

### Common Issues

#### 1. Dependencies Not Found
```
Error: Neither webrtcvad nor pydub is available
```
**Solution:** Install the required packages:
```bash
pip install webrtcvad pydub
```

#### 2. WebRTC VAD Not Working
```
Error: webrtcvad is not available
```
**Solution:** Install webrtcvad or use pydub method:
```python
detector = create_silence_detector(method='pydub')
```

#### 3. Audio Sample Rate Issues
```
Warning: Sample rate 44100 not supported by webrtcvad, using 16000
```
**Solution:** This is handled automatically, but you can set the sample rate explicitly:
```python
detector = create_silence_detector(sample_rate=16000)
```

### Performance Tips

1. **Choose the right method:**
   - Use `webrtcvad` for real-time speech detection
   - Use `pydub` for general audio analysis
   - Use `auto` for automatic selection

2. **Adjust parameters for your environment:**
   - **Noisy environments:** Increase `min_silence_len`, decrease `silence_thresh`
   - **Quiet environments:** Decrease `min_silence_len`, increase `silence_thresh`

3. **Use appropriate presets:**
   - `sensitive` for quiet environments
   - `balanced` for general use
   - `aggressive` for noisy environments

## Testing

Run the demonstration script to test the silence detection features:

```bash
# Show all features
python demo_silence_detection.py

# Show specific features
python demo_silence_detection.py --cli
python demo_silence_detection.py --programmatic
python demo_silence_detection.py --realtime
python demo_silence_detection.py --gui
python demo_silence_detection.py --integration
```

## Contributing

When contributing to the silence detection system:

1. **Test with both detection methods** (webrtcvad and pydub)
2. **Validate across different audio environments**
3. **Ensure backward compatibility** with existing code
4. **Update documentation** for new features
5. **Add tests** for new functionality

## License

This silence detection system is part of the Fluent AI project and follows the same license terms.
