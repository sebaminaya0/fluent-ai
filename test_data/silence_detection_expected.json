{
  "short_silence.wav": {
    "description": "Audio with 500ms silence in middle",
    "expected_silences": [
      {
        "start": 1.0,
        "duration": 0.5,
        "end": 1.5
      }
    ]
  },
  "long_silence.wav": {
    "description": "Audio with 1.5s silence in middle",
    "expected_silences": [
      {
        "start": 1.0,
        "duration": 1.5,
        "end": 2.5
      }
    ]
  },
  "multiple_silences.wav": {
    "description": "Audio with multiple silence periods",
    "expected_silences": [
      {
        "start": 0.5,
        "duration": 0.8,
        "end": 1.3
      },
      {
        "start": 1.8,
        "duration": 1.2,
        "end": 3.0
      }
    ]
  },
  "no_silence.wav": {
    "description": "Continuous speech with no significant silence",
    "expected_silences": []
  },
  "noisy_silence.wav": {
    "description": "Silence with background noise",
    "expected_silences": [
      {
        "start": 1.0,
        "duration": 1.0,
        "end": 2.0
      }
    ]
  },
  "very_quiet_speech.wav": {
    "description": "Very quiet speech that might be detected as silence",
    "expected_silences": []
  }
}